src/file/writer.rs

```diff
@@ -0,0 +1,447 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::io::Write;
+
+use basic::PageType;
+use column::page::{CompressedPage, Page, PageWriter};
+use errors::Result;
+use file::metadata::ColumnChunkMetaData;
+use file::statistics::{to_thrift as statistics_to_thrift};
+use parquet_format as parquet;
+use thrift::protocol::{TCompactOutputProtocol, TOutputProtocol};
+use util::io::Position;
+
+/// Serialized page writer.
+///
+/// Writes and serializes pages and metadata into output stream.
+/// `SerializedPageWriter` should not be used after calling `close()`, metrics
+/// (offsets and sizes) are not reset, create a new page writer instead.
+pub struct SerializedPageWriter<T: Write + Position> {
+  sink: T,
+  // Dictionary offset is set only once, panics if another attempt is made.
+  dictionary_page_offset: Option<u64>,
+  // Data page offset is set once with the first page.
+  data_page_offset: Option<u64>,
+  total_uncompressed_size: u64,
+  total_compressed_size: u64,
+  num_values: u32
+}
+
+impl<T: Write + Position> SerializedPageWriter<T> {
+  /// Creates new page writer.
+  pub fn new(sink: T) -> Self {
+    Self {
+      sink: sink,
+      dictionary_page_offset: None,
+      data_page_offset: None,
+      total_uncompressed_size: 0,
+      total_compressed_size: 0,
+      num_values: 0
+    }
+  }
+
+  /// Serializes page header into Thrift.
+  #[inline]
+  fn serialize_page_header(&mut self, header: parquet::PageHeader) -> Result<usize> {
+    let start_pos = self.sink.pos();
+    {
+      let mut protocol = TCompactOutputProtocol::new(&mut self.sink);
+      header.write_to_out_protocol(&mut protocol)?;
+      protocol.flush()?;
+    }
+    Ok((self.sink.pos() - start_pos) as usize)
+  }
+
+  /// Serializes column chunk into Thrift.
+  #[inline]
+  fn serialize_column_chunk(&mut self, chunk: parquet::ColumnChunk) -> Result<()> {
```

Comment:
Just for my understanding, this will only be called once per PageWriter instance, is that right?

Reply:
Yes, it will only be called when we finalise column writer. I know it is a bit confusing, it could be part of a column writer.